File: java/src/main/java/it/giannibombelli/wsc2026/payment/domain/ports/PaymentRepository.java
package it.giannibombelli.wsc2026.payment.domain.ports;

import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;

import java.util.List;
import java.util.Optional;

public interface PaymentRepository {
    void save(Payment payment);

    Optional<Payment> findById(PaymentId id);

    Optional<Payment> findByClientReference(ClientReference clientReference);

    List<Payment> findAllRequestedAndProcessingBefore(Timestamp threshold);
}

File: typescript/src/payment/domain/ports/paymentRepository.ts
// Porta per la persistenza degli aggregati Payment.

import type { ClientReference } from '@/common/domain/primitive/clientReference.js';
import type { Timestamp } from '@/common/domain/primitive/timestamp.js';
import type { Payment } from '../payment/payment.js';
import type { PaymentId } from '../payment/paymentId.js';

export interface PaymentRepository {
  save(payment: Payment): void;

  findById(id: PaymentId): Payment | null;

  findByClientReference(clientReference: ClientReference): Payment | null;

  findAllRequestedAndProcessingBefore(threshold: Timestamp): Payment[];
}
